controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java

```diff
@@ -0,0 +1,119 @@
+import java.util.HashSet;
+import java.util.OptionalInt;
+import java.util.function.Predicate;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Updates refreshed endpoint certificates and triggers redeployment, and deletes unused certificates
+ *
+ * @author andreer
+ */
+public class EndpointCertificateMaintainer extends ControllerMaintainer {
+
+    private static final Logger log = Logger.getLogger(EndpointCertificateMaintainer.class.getName());
+
+    private final DeploymentTrigger deploymentTrigger;
+    private final Clock clock;
+    private final CuratorDb curator;
+    private final SecretStore secretStore;
+    private final EndpointCertificateProvider endpointCertificateProvider;
+    private final BooleanFlag useEndpointCertificateMaintainer;
+
+    public EndpointCertificateMaintainer(Controller controller, Duration interval, Clock clock, FlagSource flagSource, SecretStore secretStore) {
+        super(controller, interval, null, SystemName.allOf(Predicate.not(SystemName::isPublic)));
+        this.deploymentTrigger = controller.applications().deploymentTrigger();
+        this.clock = clock;
+        this.secretStore = secretStore;
+        this.curator = controller().curator();
+        this.endpointCertificateProvider = controller.serviceRegistry().endpointCertificateProvider();
+        this.useEndpointCertificateMaintainer = Flags.USE_ENDPOINT_CERTIFICATE_MAINTAINER.bindTo(flagSource);
+    }
+
+    @Override
+    protected boolean maintain() {
+
+        if(!useEndpointCertificateMaintainer.value())
+            return true; // handled by EndpointCertificateManager for now
+
+        try {
+            deleteUnusedCertificates();
+            updateRefreshedCertificates();
+        } catch (Exception e) {
+            log.log(LogLevel.WARNING, "failed lol", e);
+            return false;
+        }
+
+        return true;
+    }
+
+    private void updateRefreshedCertificates() {
+        curator.readAllEndpointCertificateMetadata().forEach(((applicationId, endpointCertificateMetadata) -> {
+            // Look for and use refreshed certificate
+            var latestAvailableVersion = latestVersionInSecretStore(endpointCertificateMetadata);
+            if (latestAvailableVersion.isPresent() && latestAvailableVersion.getAsInt() > endpointCertificateMetadata.version()) {
+                var refreshedCertificateMetadata = endpointCertificateMetadata.withVersion(latestAvailableVersion.getAsInt());
+                curator.writeEndpointCertificateMetadata(applicationId, refreshedCertificateMetadata);
+                // TODO andreer: Certificate not validated here! Is that OK?
+                deploymentTrigger.triggerChange(applicationId, Change.empty()); // TODO andreer: No idea if this does what I want, or anything at all
+            }
+        }));
+    }
+
+    private OptionalInt latestVersionInSecretStore(EndpointCertificateMetadata originalCertificateMetadata) {
```

Comment:
Do we somehow want to catch (report/alert) situations where certversion<keyversion for a long time?

Reply:
It should cause an alert to oncall after a week, as the cert nears expiry. I think that's good enough